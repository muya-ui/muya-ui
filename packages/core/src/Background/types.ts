import React from 'react';

import { IBackgroundColor } from '@muya-ui/theme-light';

export interface IBackgroundPureProps {
  /**
   * 支持通用背景的几种模式
   * @default 'global'
   */
  type?: keyof IBackgroundColor;
  /**
   * 节点设置
   * @default 'div'
   */
  component?: 'div' | 'span';
  /** css display */
  display?: 'block' | 'flex' | 'inline-block' | 'inline-flex' | React.CSSProperties['display'];
  /**
   * 设置具体的背景色，酌情使用
   */
  backgroundColor?: string;
  /**
   * 文字颜色，酌情使用
   */
  color?: string;
  /** 宽 */
  width?: string | number;
  /** 高 */
  height?: string | number;
}

export type IBackgroundProps = IBackgroundPureProps &
  React.HTMLAttributes<HTMLDivElement | HTMLSpanElement>;
