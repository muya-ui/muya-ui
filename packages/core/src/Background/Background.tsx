import React from 'react';
import styled, { css } from 'styled-components';

import { IThemedBaseProps } from '../types';
import addPx from '../utils/addPx';
import memoForwardRef from '../utils/memoForwardRef';
import useTheme from '../utils/useTheme';
import { IBackgroundProps } from './types';
import warning from 'warning';

interface IRootProps {
  $type?: IBackgroundProps['type'];
  $display?: React.CSSProperties['display'];
  $backgroundColor?: string;
  $color?: string;
  $height?: string | number;
  $width?: string | number;
}

function baseCSS({
  $display,
  $width,
  $backgroundColor,
  $color,
  $height,
  $type = 'global',
  theme,
}: IRootProps & IThemedBaseProps) {
  let backgroundColor;
  if ($backgroundColor) {
    backgroundColor = $backgroundColor;
  } else {
    backgroundColor = theme.colors.pattern.background[$type];
  }
  const color = $color || theme.colors.pattern.text.text;
  return css({
    display: $display,
    backgroundColor,
    color,
    width: addPx($width),
    height: addPx($height),
  });
}

const RootDiv = styled.div<IRootProps>`
  ${baseCSS}
`;
const RootSpan = styled.div<IRootProps>`
  ${baseCSS}
`;

const Background = memoForwardRef<HTMLDivElement | HTMLSpanElement, IBackgroundProps>(
  (props, ref) => {
    const { type, component, display, backgroundColor, width, height, color, ...other } = props;
    const theme = useTheme();
    const Root = component === 'span' ? RootSpan : RootDiv;
    warning(
      !(type === 'scrollBar' || type === 'scrollBarHover'),
      '[Background]: type scrollBar and scrollBarHover is deprecated, will removed in 0.5.0',
    );
    return (
      <Root
        {...other}
        $type={type}
        $color={color}
        $display={display}
        $width={width}
        $height={height}
        $backgroundColor={backgroundColor}
        theme={theme}
        ref={ref as any}
      />
    );
  },
);

export default Background;
