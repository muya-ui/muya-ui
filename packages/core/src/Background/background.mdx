---
name: Background 背景
menu: 组件
route: /components/background
autoDemos: true
parent: 通用
disableDoubleRow: false
---


# Background 背景

背景组件。

{Demos}

### API

#### IBackgroundProps

<Props propsName="IBackgroundProps" />
