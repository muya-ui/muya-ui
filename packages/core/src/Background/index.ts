export * from './types';
export { default } from './Background';
export { default as Background } from './Background';
