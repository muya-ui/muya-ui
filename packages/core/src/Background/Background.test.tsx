import React from 'react';
import renderer from 'react-test-renderer';

import Background from './Background';

test('Background 正常情况', () => {
  const tree = renderer.create(<Background />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Background 不同 type', () => {
  const tree = renderer
    .create(
      <div>
        <Background type="global">背景模式：global</Background>
        <Background type="higher">背景模式：higher</Background>
        <Background type="block">背景模式：block</Background>
        <Background type="selectedBlock">背景模式：selectedBlock</Background>
        <Background type="disabled">背景模式：disabled</Background>
        <Background type="divider">背景模式：divider</Background>
        <Background type="mask" color="#FFF">
          背景模式：mask
        </Background>
      </div>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Background 设置宽高', () => {
  const tree = renderer.create(<Background width={100} height={100} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Background 设置颜色', () => {
  const tree = renderer.create(<Background color="#FFF" backgroundColor="#000" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Background 设置 display', () => {
  const tree = renderer.create(<Background display="inline-block" component="span" />).toJSON();
  expect(tree).toMatchSnapshot();
});
