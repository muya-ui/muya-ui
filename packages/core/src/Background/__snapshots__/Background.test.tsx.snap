// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background 不同 type 1`] = `
.c0 {
  background-color: rgba(255,255,255,1);
  color: rgba(102,102,102,1);
}

.c1 {
  background-color: rgba(250,250,250,1);
  color: rgba(102,102,102,1);
}

.c2 {
  background-color: rgba(238,238,238,1);
  color: rgba(102,102,102,1);
}

.c3 {
  background-color: rgba(0,0,0,1);
  color: #FFF;
}

<div>
  <div
    className="c0"
  >
    背景模式：global
  </div>
  <div
    className="c0"
  >
    背景模式：higher
  </div>
  <div
    className="c0"
  >
    背景模式：block
  </div>
  <div
    className="c0"
  >
    背景模式：selectedBlock
  </div>
  <div
    className="c1"
  >
    背景模式：disabled
  </div>
  <div
    className="c2"
  >
    背景模式：divider
  </div>
  <div
    className="c3"
  >
    背景模式：mask
  </div>
</div>
`;

exports[`Background 正常情况 1`] = `
.c0 {
  background-color: rgba(255,255,255,1);
  color: rgba(102,102,102,1);
}

<div
  className="c0"
/>
`;

exports[`Background 设置 display 1`] = `
.c0 {
  display: inline-block;
  background-color: rgba(255,255,255,1);
  color: rgba(102,102,102,1);
}

<div
  className="c0"
/>
`;

exports[`Background 设置宽高 1`] = `
.c0 {
  background-color: rgba(255,255,255,1);
  color: rgba(102,102,102,1);
  width: 100px;
  height: 100px;
}

<div
  className="c0"
/>
`;

exports[`Background 设置颜色 1`] = `
.c0 {
  background-color: #000;
  color: #FFF;
}

<div
  className="c0"
/>
`;
